lib/std/enums.zig

```diff
@@ -1415,3 +1415,42 @@ test "std.enums.EnumIndexer sparse" {
     try testing.expectEqual(E.b, Indexer.keyForIndex(1));
     try testing.expectEqual(E.c, Indexer.keyForIndex(2));
 }
+
+/// Returns the variant of an enum type, `T`, which is named `str`, or `null` if no such variant exists.
+pub fn fromString(comptime T: type, str: []const u8) ?T {
+    // Using ComptimeStringMap here is more performant, but it will start to take too
+    // long to compile if the enum is large enough, due to the current limits of comptime
+    // performance when doing things like constructing lookup maps at comptime.
+    // TODO The '100' here is arbitrary and should be increased when possible:
+    // - https://github.com/ziglang/zig/issues/4055
+    // - https://github.com/ziglang/zig/issues/3863
+    if (@typeInfo(T).Enum.fields.len <= 100) {
+        const kvs = comptime build_kvs: {
+            const EnumKV = struct { []const u8, T };
+            var kvs_array: [@typeInfo(T).Enum.fields.len]EnumKV = undefined;
+            inline for (@typeInfo(T).Enum.fields, 0..) |enumField, i| {
+                kvs_array[i] = .{ enumField.name, @field(T, enumField.name) };
+            }
+            break :build_kvs kvs_array[0..];
+        };
+        const map = std.ComptimeStringMap(T, kvs);
+        return map.get(str);
+    } else {
+        inline for (@typeInfo(T).Enum.fields) |enumField| {
+            if (std.mem.eql(u8, str, enumField.name)) {
+                return @field(T, enumField.name);
+            }
+        }
+        return null;
+    }
+}
```

Comment:
If this can't be done in a more efficient manner, then there is a strong case to be made for this function to become a builtin. In that case, it would be a second breaking change. So, I think we should wait to move this function until that question becomes answered.